File: Python/Password_Generator.py
import secrets
import string


def generate_password(length: int = 16) -> str:
    if length < 8:
        raise ValueError("Password length should be at least 8.")

    lowercase = string.ascii_lowercase
    uppercase = string.ascii_uppercase
    digits = string.digits
    symbols = "!@#$%^&*()-_=+[]{};:,.?"
    all_chars = lowercase + uppercase + digits + symbols

    # Ensure at least one char from each group
    required = [
        secrets.choice(lowercase),
        secrets.choice(uppercase),
        secrets.choice(digits),
        secrets.choice(symbols),
    ]

    remaining = [secrets.choice(all_chars) for _ in range(length - len(required))]
    password_chars = required + remaining

    # Secure shuffle
    secrets.SystemRandom().shuffle(password_chars)
    return "".join(password_chars)


def generate_unique_passwords(count: int, length: int) -> list[str]:
    passwords = set()

    while len(passwords) < count:
        passwords.add(generate_password(length))

    return list(passwords)


def main() -> None:
    try:
        count = int(input("How many passwords? ").strip() or "1")
        length = int(input("Password length? ").strip() or "16")

        if count < 1:
            print("Count must be at least 1.")
            return

        passwords = generate_unique_passwords(count, length)

        print("\nGenerated unique passwords:")
        for index, password in enumerate(passwords, start=1):
            print(f"{index}. {password}")

    except ValueError as error:
        print(f"Error: {error}")


if __name__ == "__main__":
    main()

File: Typescript/Password_Generator.ts
const crypto = require('node:crypto');
const readline = require('node:readline/promises');
const { stdin, stdout } = require('node:process');

const lowercase = 'abcdefghijklmnopqrstuvwxyz';
const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const digits = '0123456789';
const symbols = '!@#$%^&*()-_=+[]{};:,.?';
const allChars = lowercase + uppercase + digits + symbols;

const randomChar = (pool) => pool[crypto.randomInt(0, pool.length)];

const shuffle = (items) => {
  for (let i = items.length - 1; i > 0; i -= 1) {
    const j = crypto.randomInt(0, i + 1);
    [items[i], items[j]] = [items[j], items[i]];
  }
};

const generatePassword = (length = 16) => {
  if (length < 8) {
    throw new Error('Password length should be at least 8.');
  }

  const required = [
    randomChar(lowercase),
    randomChar(uppercase),
    randomChar(digits),
    randomChar(symbols),
  ];

  const remaining = Array.from(
    { length: length - required.length },
    () => randomChar(allChars),
  );

  const passwordChars = [...required, ...remaining];
  shuffle(passwordChars);
  return passwordChars.join('');
};

const generateUniquePasswords = (count, length) => {
  const passwords = new Set();
  while (passwords.size < count) {
    passwords.add(generatePassword(length));
  }
  return [...passwords];
};

const main = async () => {
  const rl = readline.createInterface({ input: stdin, output: stdout });

  try {
    const count = Number((await rl.question('How many passwords? ')).trim() || '1');
    const length = Number((await rl.question('Password length? ')).trim() || '16');

    if (!Number.isInteger(count) || count < 1) {
      console.log('Count must be at least 1.');
      return;
    }

    const passwords = generateUniquePasswords(count, length);

    console.log('\nGenerated unique passwords:');
    passwords.forEach((password, index) => {
      console.log(`${index + 1}. ${password}`);
    });
  } catch (error) {
    console.log(`Error: ${error.message}`);
  } finally {
    rl.close();
  }
};

main();
